File: apps/backend/billingms/src/test/java/com/example/billingms/domain/model/MoneyTest.java
package com.example.billingms.domain.model;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.math.BigDecimal;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

/**
 * 金額（[ADR-027] 決定 2）。
 *
 * <p><strong>丸めはここ 1 か所だけで行う。</strong>呼び出し側が丸めると、丸める場所が
 * 増えるたびに結果が変わりうる——画面と保存値が食い違い、明細の合計と総額が 1 円ずれる。
 */
@DisplayName("金額")
class MoneyTest {

    @Nested
    @DisplayName("丸め")
    class Rounding {

        /**
         * <strong>1 円単位で四捨五入する</strong>（決定 2）。
         *
         * <p>割引後の金額には端数が出る（基本料金 × 0.85 など）。
         */
        @Test
        @DisplayName("端数は 1 円単位で四捨五入する")
        void roundsToTheNearestYen() {
            assertThat(Money.yen(new BigDecimal("100.4")).amount())
                    .isEqualByComparingTo("100");
            assertThat(Money.yen(new BigDecimal("100.5")).amount())
                    .isEqualByComparingTo("101");
            assertThat(Money.yen(new BigDecimal("100.6")).amount())
                    .isEqualByComparingTo("101");
        }

        /**
         * <strong>負の金額も同じ向きで丸める。</strong>
         *
         * <p>減額（明細）は負の数で入る。HALF_UP は 0 から遠いほうへ丸めるため、
         * -100.5 は -101 になる。<strong>「金額の絶対値で見て同じ扱い」</strong>であり、
         * 減額だけ有利／不利に倒れない。
         */
        @Test
        @DisplayName("負の金額も絶対値で見て同じ向きに丸める")
        void roundsNegativeAmountsSymmetrically() {
            assertThat(Money.yen(new BigDecimal("-100.5")).amount())
                    .isEqualByComparingTo("-101");
            assertThat(Money.yen(new BigDecimal("-100.4")).amount())
                    .isEqualByComparingTo("-100");
        }

        /** <strong>丸めた結果を保持する。</strong>あとから丸め直せる形で持たない。 */
        @Test
        @DisplayName("保持している値は丸めたあとの値である")
        void keepsTheRoundedValue() {
            Money money = Money.yen(new BigDecimal("420000.7"));

            assertThat(money.amount().scale())
                    .as("小数を抱えたままだと、表示のたびに丸め直すことになる")
                    .isLessThanOrEqualTo(0);
        }
    }

    @Nested
    @DisplayName("演算")
    class Arithmetic {

        /** 掛け算の結果も丸める。**掛けてから丸めるのであって、丸めてから掛けるのではない。** */
        @Test
        @DisplayName("掛け算の結果は丸めて返す")
        void multipliesAndRounds() {
            Money base = Money.yen(new BigDecimal("420000"));

            assertThat(base.multiply(new BigDecimal("0.085")).amount())
                    .isEqualByComparingTo("35700");
        }

        @Test
        @DisplayName("足し算と引き算ができる")
        void addsAndSubtracts() {
            Money base = Money.yen(new BigDecimal("420000"));

            assertThat(base.add(Money.yen(new BigDecimal("10000"))).amount())
                    .isEqualByComparingTo("430000");
            assertThat(base.subtract(Money.yen(new BigDecimal("42000"))).amount())
                    .isEqualByComparingTo("378000");
        }

        /**
         * <strong>通貨が違う金額は足せない。</strong>
         *
         * <p>足せてしまうと、円とドルが混ざった請求書ができる。多通貨は将来の話だが、
         * <strong>混ぜられる形にしておくと、混ざったことに気づけない</strong>。
         */
        @Test
        @DisplayName("通貨の違う金額は足せない")
        void rejectsMixedCurrencies() {
            Money yen = Money.yen(new BigDecimal("1000"));
            Money dollar = Money.of(new BigDecimal("1000"), "USD");

            assertThatThrownBy(() -> yen.add(dollar))
                    .isInstanceOf(IllegalArgumentException.class)
                    .hasMessageContaining("通貨");
        }
    }

    @Nested
    @DisplayName("生成")
    class Creation {

        @Test
        @DisplayName("通貨コードは 3 文字である")
        void requiresAThreeLetterCurrencyCode() {
            assertThatThrownBy(() -> Money.of(BigDecimal.ONE, "JPYEN"))
                    .isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> Money.of(BigDecimal.ONE, null))
                    .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        @DisplayName("金額が無い状態は作れない")
        void requiresAnAmount() {
            assertThatThrownBy(() -> Money.of(null, "JPY"))
                    .isInstanceOf(IllegalArgumentException.class);
        }

        /** **0 円は作れる。** 調整で相殺されて 0 になることはある。 */
        @Test
        @DisplayName("0 円は作れる")
        void allowsZero() {
            assertThat(Money.yen(BigDecimal.ZERO).amount()).isEqualByComparingTo("0");
        }
    }

    @Nested
    @DisplayName("符号と表示")
    class SignAndText {

        /** 減額（明細）は負の数で入る。**符号で判別できないと、減額が加算に化ける。** */
        @Test
        @DisplayName("負の金額を判別できる")
        void tellsNegativeAmounts() {
            assertThat(Money.yen(new BigDecimal("-1")).isNegative()).isTrue();
            assertThat(Money.yen(BigDecimal.ZERO).isNegative()).isFalse();
            assertThat(Money.yen(new BigDecimal("1")).isNegative()).isFalse();
        }

        @Test
        @DisplayName("0 円を作れる")
        void createsZero() {
            assertThat(Money.zero()).isEqualTo(Money.yen(BigDecimal.ZERO));
            assertThat(Money.zero().currency()).isEqualTo("JPY");
        }

        /** ログや例外の文面に出る。**金額だけだと、通貨が分からない。** */
        @Test
        @DisplayName("金額と通貨を並べて表す")
        void showsAmountAndCurrency() {
            assertThat(Money.yen(new BigDecimal("1000"))).hasToString("1000 JPY");
        }

        /** **同じ金額はハッシュも同じ。** 集合やマップの鍵にしたときに壊れない。 */
        @Test
        @DisplayName("等しい金額はハッシュも等しい")
        void hashesConsistently() {
            assertThat(Money.yen(new BigDecimal("1000")).hashCode())
                    .isEqualTo(Money.yen(new BigDecimal("1000.0")).hashCode());
        }

        @Test
        @DisplayName("金額でないものとは等しくない")
        void isNotEqualToOtherTypes() {
            assertThat(Money.yen(BigDecimal.ONE)).isNotEqualTo("1 JPY").isNotEqualTo(null);
            assertThat(Money.yen(BigDecimal.ONE)).isEqualTo(Money.yen(BigDecimal.ONE));
        }
    }

    @Nested
    @DisplayName("引き算と掛け算の拒み方")
    class InvalidArithmetic {

        @Test
        @DisplayName("通貨の違う金額は引けない")
        void rejectsMixedCurrenciesOnSubtract() {
            Money yen = Money.yen(new BigDecimal("1000"));
            Money dollar = Money.of(new BigDecimal("1000"), "USD");

            assertThatThrownBy(() -> yen.subtract(dollar))
                    .isInstanceOf(IllegalArgumentException.class)
                    .hasMessageContaining("通貨");
        }

        @Test
        @DisplayName("相手が無ければ計算できない")
        void rejectsMissingOperand() {
            Money yen = Money.yen(new BigDecimal("1000"));

            assertThatThrownBy(() -> yen.add(null))
                    .isInstanceOf(IllegalArgumentException.class);
            assertThatThrownBy(() -> yen.multiply(null))
                    .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    @DisplayName("等価性")
    class Equality {

        /** **同じ金額は同じ値である。** 値オブジェクトとして比較できないと、検査が書けない。 */
        @Test
        @DisplayName("同じ金額と通貨なら等しい")
        void comparesByValue() {
            assertThat(Money.yen(new BigDecimal("1000")))
                    .isEqualTo(Money.yen(new BigDecimal("1000")));
            assertThat(Money.yen(new BigDecimal("1000")))
                    .isNotEqualTo(Money.of(new BigDecimal("1000"), "USD"));
        }

        /**
         * <strong>丸める前が違っても、丸めた結果が同じなら等しい。</strong>
         *
         * <p>100.4 円と 100.4999 円はどちらも 100 円である。等価性が丸め前に依存すると、
         * 同じ請求金額なのに「違う」と判定される。
         */
        @Test
        @DisplayName("丸めた結果が同じなら等しい")
        void comparesAfterRounding() {
            assertThat(Money.yen(new BigDecimal("100.4")))
                    .isEqualTo(Money.yen(new BigDecimal("100.49")));
        }
    }
}

File: docs/article/source/java-3/apps/frontend/src/features/billing/__tests__/money.test.ts
import { describe, expect, it } from 'vitest'

import { formatRate, formatYen, roundYen } from '../money'

/**
 * 金額と率の整形（[ADR-027] 決定 2）。
 *
 * <p><strong>ここでは計算しない。</strong>丸めはサーバが済ませており、画面は返ってきた
 * 値をそのまま出す。フロントで丸めると丸める場所が 2 か所になり、画面と保存値が食い違う。
 */
describe('金額の表示', () => {
  it('桁区切りを入れて円で出す', () => {
    expect(formatYen({ value: 1_260_000, currency: 'JPY' })).toBe('¥1,260,000')
  })

  /** 減額は負の数で入る（決定 6）。**符号を落とすと、減額が加算に見える。** */
  it('負の金額は符号を先に出す', () => {
    expect(formatYen({ value: -10_000, currency: 'JPY' })).toBe('-¥10,000')
  })

  /** **無いことと 0 は違う。** キャンセル料が無い予約に ¥0 を出すと、0 円が算定されたと読める。 */
  it('金額が無ければ「—」を出す', () => {
    expect(formatYen(null)).toBe('—')
    expect(formatYen(undefined)).toBe('—')
  })

  it('0 円は 0 円として出す', () => {
    expect(formatYen({ value: 0, currency: 'JPY' })).toBe('¥0')
  })
})

/**
 * IT11 レビュー 高 2。**サーバと同じ向きに丸める。**
 *
 * <p>画面が `Math.round` を使うと、サーバの `HALF_UP`（0 から遠いほうへ丸める）と
 * 向きが違う——**小計が負になる調整**（大幅な減額・補償）で 1 円ずれる。
 * 金額を扱う画面で最も信頼を失う種類の食い違いであり、担当者は「どちらが本当の
 * 請求額か」を確かめる手段を持たない。
 */
describe('円への丸め', () => {
  it('正の端数は四捨五入する', () => {
    expect(roundYen(100.4)).toBe(100)
    expect(roundYen(100.5)).toBe(101)
    expect(roundYen(100.6)).toBe(101)
  })

  /**
   * **負の端数は 0 から遠いほうへ丸める**（サーバの HALF_UP と同じ）。
   *
   * `Math.round(-1.5)` は -1 を返す（+∞ 方向）。サーバは -2 を返す。
   */
  it('負の端数は 0 から遠いほうへ丸める', () => {
    expect(roundYen(-1.5), 'Math.round のままだと -1 になり、サーバと 1 円ずれる').toBe(-2)
    expect(roundYen(-2.5)).toBe(-3)
    expect(roundYen(-1.4)).toBe(-1)
    expect(roundYen(-1.6)).toBe(-2)
  })

  it('端数が無ければそのまま', () => {
    expect(roundYen(0)).toBe(0)
    expect(roundYen(-100)).toBe(-100)
    expect(roundYen(378000)).toBe(378000)
  })
})

describe('率の表示', () => {
  /** **率そのものを出す**（22-4）。額だけでは率を復元できない——割り戻すと丸めの分ずれる。 */
  it('割合を百分率で出す', () => {
    expect(formatRate(0.1)).toBe('10%')
    expect(formatRate(0.3)).toBe('30%')
  })

  it('端数のある率は小数第 1 位まで出す', () => {
    expect(formatRate(0.155)).toBe('15.5%')
  })

  /**
   * **契約が無いことと 0% は違う**（[ADR-012] と同じ判断）。
   *
   * 0% を出すと「割引が 0 だった」に読め、契約が無いことと区別できない。
   */
  it('率が無ければ「—」を出す', () => {
    expect(formatRate(null)).toBe('—')
  })

  it('0% は 0% として出す', () => {
    expect(formatRate(0)).toBe('0%')
  })
})
